command/013_config_upgrade.go

```diff
@@ -71,58 +75,386 @@ func (c *ZeroThirteenUpgradeCommand) Run(args []string) int {
+					diags = diags.Append(&hcl.Diagnostic{
+						Summary:  "Duplicate required provider configuration",
+						Detail:   fmt.Sprintf("Found duplicate required provider configuration for %q.Previously configured at %s", rp.Name, previous.DeclRange),
+						Severity: hcl.DiagWarning,
+						Context:  rps.DeclRange.Ptr(),
+						Subject:  rp.DeclRange.Ptr(),
+					})
+				} else {
+					// We're copying the struct here to ensure that any
+					// mutation does not affect the original, if we rewrite
+					// this file
+					requiredProviders[rp.Name] = &configs.RequiredProvider{
+						Name:        rp.Name,
+						Source:      rp.Source,
+						Type:        rp.Type,
+						Requirement: rp.Requirement,
+						DeclRange:   rp.DeclRange,
+					}
+				}
+			}
+		}
+	}
+
+	for _, file := range files {
+		// Step 2: add missing provider requirements from provider blocks
+		for _, p := range file.ProviderConfigs {
+			// If no explicit provider configuration exists for the
+			// provider configuration's local name, add one with a legacy
+			// provider address.
+			if _, exist := requiredProviders[p.Name]; !exist {
+				requiredProviders[p.Name] = &configs.RequiredProvider{
+					Name:        p.Name,
+					Type:        addrs.NewLegacyProvider(p.Name),
+					Requirement: p.Version,
+				}
+			}
+		}
+
+		// Step 3: add missing provider requirements from resources
+		resources := [][]*configs.Resource{file.ManagedResources, file.DataResources}
+		for _, rs := range resources {
+			for _, r := range rs {
+				// Find the appropriate provider local name for this resource
+				var localName string
+
+				// If there's a provider config, use that to determine the
+				// local name. Otherwise use the implied provider local name
+				// based on the resource's address.
+				if r.ProviderConfigRef != nil {
+					localName = r.ProviderConfigRef.Name
+				} else {
+					localName = r.Addr().ImpliedProvider()
+				}
+
+				// If no explicit provider configuration exists for this local
+				// name, add one with a legacy provider address.
+				if _, exist := requiredProviders[localName]; !exist {
+					requiredProviders[localName] = &configs.RequiredProvider{
+						Name: localName,
+						Type: addrs.NewLegacyProvider(localName),
+					}
+				}
+			}
+		}
 	}
 
-	// Generate the required providers configuration
-	genDiags := generateRequiredProviders(providerSources, dir)
-	diags = diags.Append(genDiags)
+	// We should now have a complete understanding of the provider requirements
+	// stated in the config.  If there are any providers, attempt to detect
+	// their sources, and rewrite the config.
+	if len(requiredProviders) > 0 {
+		detectDiags := c.detectProviderSources(requiredProviders)
+		diags = diags.Append(detectDiags)
+		if diags.HasErrors() {
+			c.Ui.Error("Unable to detect sources for providers")
+			c.showDiagnostics(diags)
+			return 1
+		}
+
+		// Default output filename is "providers.tf"
+		filename := "providers.tf"
```

Comment:
I do like the idea of updating the required version (and will revisit that!), but I'd be surprised to see `required_providers` blocks with _source_ fields in a `versions.tf` file. I think the `providers.tf` name (if there is no existing `required_providers` block) is a more obvious name for the purpose.

Reply:
We named this file `versions.tf` in anticipation of it conventionally being the one place to document which versions of things you were depending on, rather than this being spread all over the module code, but it is true that the addition of `source` has made this a little more complicated.

I think I don't feel so weird about it because I think perhaps my mental model is a bit different. For example:

```hcl
terraform {
  required_providers {
    aws = { source = "hashicorp/aws", version = "2.0.0" }
  }
}
```

I read the above as "this module needs version 2.0.0 of `hashicorp/aws` and will refer to it elsewhere in this module as `aws`". I can see an alternative reading of this as "This module needs 2.0.0 of the `aws` provider and the location from which to fetch it is `hashicorp/aws`", which makes this feel less like a pure version constraint.

It's kinda unfortunate in retrospect that we named this attribute `source` because it makes it sound like it's just a location rather than being the primary identifier for this provider, but I wasn't able to come up with any better name for it and so I've just kinda accepted it. :upside_down_face: 

With that said, I don't feel super strongly about this suggestion, so I'm happy for it to stay in `providers.tf` if that feels clearer to you. I kinda wish then that we'd called the other file `version.tf` (singular), but that's not a big deal. :grinning:

Reply:
That's a useful explanation, thanks! I'm now inclined towards `versions.tf` over `providers.tf`, and will look into making that change as part of the `required_version` update work.